internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1068,122 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkOldSnapshotsPipelineRun sorts all snapshots for application and cancells all running interationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as cancelled run finally to be gracefuly cancelled
+// god please forgive me for this monstrosity of cognitive complexity 40+
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+	snapshots, err := a.loader.GetAllSnapshots(a.context, a.client, application)
```

Comment:
I think it would be a good optimization here to add a new function to loader which gets all Snapshots for a given PR, similar to how it's done in the [GetAllSnapshotsForBuildPipelineRun](https://github.com/konflux-ci/integration-service/blob/9b2d7160579f1a24bf20dc186368078916af03de/loader/loader.go#L334-L350) function, just with different labels. This way we wouldn't have to loop through all of the users Snapshots.

Reply:
Agreed, getting all snapshots is not efficient way here. Created new method just for this purpose.